### Forma de pensar

In [ ]:
# Top -> Down

# Exemplo simples: atravessar a rua

### Exemplo de Exercício do Curso

#### 23. Faça um Programa para uma loja de tintas. O programa deverá pedir o tamanho em metros quadrados da área a ser pintada. Considere que a cobertura da tinta é de 1 litro para cada 6 metros quadrados e que a tinta é vendida em latas de 18 litros, que custam R\\$ 80,00 ou em galões de 3,6 litros, que custam R\\$ 25,00.

In [22]:
import math
def menu():
    texto_menu = f'Bem vindo!'
    print(f'{texto_menu:-^30}')

def calcular_litros(metros_quadrados: float) -> float:
    rendimento_litro = 6 # metros quadrados
    return metros_quadrados / rendimento_litro

    
def calcular_custo_area_latas(metros_quadrados: float) -> float:
    custo_latas = 80 # reais
    litros_necessarios = calcular_litros(metros_quadrados)
    tot_latas, desperdicio = calcular_quantidade_latas(litros_necessarios)
    tot_custo = tot_latas * custo_latas
    print(f"Serão necessários {tot_latas} latas, custando R${padronizar_moeda(tot_custo)} para pintar a área de {metros_quadrados}m² \nCom uma sobra de {desperdicio:.2f} litros")


def calcular_quantidade_latas(litros: float) -> tuple:
    tamanho_lata = 18 # litros
    if litros > 0:
        quantidade_latas = math.ceil(litros / tamanho_lata)
        desperdicio = (quantidade_latas * tamanho_lata) % litros
        return (quantidade_latas, desperdicio)
    else:
        return (0, litros)


def calcular_custo_area_galao(metros_quadrados: float) -> float:
    custo_galao = 25 # reais
    litros_necessarios = calcular_litros(metros_quadrados)
    tot_galoes, desperdicio = calcular_quantidade_galoes(litros_necessarios)
    tot_custo = tot_galoes * custo_galao
    print(f"Serão necessários {tot_galoes} galões, custando R${padronizar_moeda(tot_custo)} para pintar a área de {metros_quadrados}m² \nCom uma sobra de {desperdicio:.2f} litros")


def calcular_quantidade_galoes(litros: float) -> tuple:
    tamanho_galao = 3.6 # litros
    if litros > 0:
        quantidade_galoes = math.ceil(litros / tamanho_galao)
        desperdicio = (quantidade_galoes * tamanho_galao) % litros
        return (quantidade_galoes, desperdicio)
    else:
        return (0, litros)


def calcular_custo_area_eficiente(metros_quadrados: float) -> int:
    custo_galao = 25 # reais
    custo_latas = 80 # reais
    litros_necessarios = calcular_litros(metros_quadrados)
    quantidade_latas, quantidade_galoes, desperdicio = calcular_quantidade_eficiente(litros_necessarios)
    tot_custo = quantidade_galoes * custo_galao + quantidade_latas * custo_latas
    print(f"Serão necessários {quantidade_latas} latas e {quantidade_galoes} galões, custando R${padronizar_moeda(tot_custo)} para pintar a área de {metros_quadrados}m² \nCom uma sobra de {desperdicio:.2f} litros")


def calcular_quantidade_eficiente(litros: float) -> tuple:
    tamanho_galao = 3.6 # litros
    tamanho_lata = 18 # litros
    custo_galao = 25 # reais
    custo_latas = 80 # reais
    comparativo_eficiencia = math.ceil(custo_latas / custo_galao) * tamanho_galao
    if litros > 0:
        if litros > comparativo_eficiencia:
            quantidade_latas = litros // tamanho_lata
            sobra = litros % tamanho_lata
            quantidade_galoes = math.ceil(sobra / tamanho_galao)
            sobra = (quantidade_galoes * tamanho_galao) // litros
            return (quantidade_latas, quantidade_galoes, sobra)
        else:
            quantidade_galoes = math.ceil(litros / tamanho_galao)
            sobra = (quantidade_galoes * tamanho_galao) - litros
            return (0, quantidade_galoes, sobra) 
    else:
        return (0, 0, litros)


def padronizar_moeda(valor: float) -> str:
    return f"{valor:_.2f}".replace(".", ",").replace("_", ".")


def padronizar_moeda(valor: str) -> str:
    return f"{valor}".replace(".", ",").replace("_",".")


menu()
metros_quadrados = float(input("Qual a área em m²? "))

print("Usando Só latas: ")
calcular_custo_area_latas(metros_quadrados)
print("-"*20)
print("Usando Só galões: ")
calcular_custo_area_galao(metros_quadrados)
print("-"*20)
print("Usando eficiente: ")
calcular_custo_area_eficiente(metros_quadrados)
print("-"*20)

----------Bem vindo!----------
Usando Só latas: 
Serão necessários 1 latas, custando R$80 para pintar a área de 107.0m² 
Com uma sobra de 0.17 litros
--------------------
Usando Só galões: 
Serão necessários 5 galões, custando R$125 para pintar a área de 107.0m² 
Com uma sobra de 0.17 litros
--------------------
Usando eficiente: 
Serão necessários 0.0 latas e 5 galões, custando R$125,0 para pintar a área de 107.0m² 
Com uma sobra de 1.00 litros
--------------------


##### Informe ao usuário as quantidades de tinta a serem compradas e os respectivos preços em 3 situações.

Dica: lembre dos operadores // e % mostrados em exercícios anteriores<br>
Dica1: numero // 10 vai te dar como resposta a parte inteira da divisão do número por 10. Ex: 7 // 3 = 2; 10 // 3 = 3<br> 
Dica2: numero % 10 vai te dar o resto da divisão do número por 10. Ex: 7 % 3 = 1; 10 % 3 = 1; 15 % 3 = 0

##### 1. Comprar apenas latas de 18 litros: (apenas latas inteiras)

##### 2. Comprar apenas galões de 3,6 litros: (apenas galoes inteiros)

##### 3. Misturar latas e galões, de forma que o desperdício de tinta seja menor. Sempre arredonde os valores para cima, isto é, considere latas cheias.

O custo da lata é 80/18 = 4,44 R\\$/L

O custo do galão é 25/3,6 = 6,94 R\\$/L

A lata é mais econômica, então todas as latas inteiras que pudermos usar devemos comprar em latas. Se ficar faltando alguma coisa para completar devemos avaliar se é melhor comprar latas ou galões. Exemplo:

Se queremos comprar 90 litros. 5 latas dão exatamente 90 litros. Então devemos comprar tudo em latas.

Se queremos comprar 95 litros. 5 latas dão exatamente 90 litros. Então devemos comprar pelo menos 5 latas e avaliar o que falta, se estes últimos 5 litros valem mais apenas em latas ou galões.

Para os 5 litros faltantes precisamos de 2 galões que custam 50 reais no total. Ou de uma lata que custa 80 reais no total. Portanto, neste caso vale mais a pena usar 2 galões.

Se queremos comprar 107 litros. 5 latas dão exatamente 90 litros. Então devemos comprar pelo menos 5 latas e avaliar o que falta, se estes últimos 5 litros valem mais apenas em latas ou galões.

Para os 17 litros faltantes precisamos de 5 galões que custam 125 reais no total. Ou de uma lata que custa 80 reais no total. Portanto, neste caso vale mais a pena usar uma lata.

3 galões custam 75 reais, 4 galões custam 100 reais. Então, se for possível completar com até 3 galões escolhe-se galões. Qualquer quantidade maior que 3 galões, usa-se latas.

Podemos ir ao exercício: